File: android/src/main/java/com/brentvatne/receiver/AudioBecomingNoisyReceiver.java
package com.brentvatne.receiver;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.media.AudioManager;

public class AudioBecomingNoisyReceiver extends BroadcastReceiver {

    private final Context context;
    private BecomingNoisyListener listener = BecomingNoisyListener.NO_OP;

    public AudioBecomingNoisyReceiver(Context context) {
        this.context = context.getApplicationContext();
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        if (AudioManager.ACTION_AUDIO_BECOMING_NOISY.equals(intent.getAction())) {
            listener.onAudioBecomingNoisy();
        }
    }

    public void setListener(BecomingNoisyListener listener) {
        this.listener = listener;
        IntentFilter intentFilter = new IntentFilter(AudioManager.ACTION_AUDIO_BECOMING_NOISY);
        context.registerReceiver(this, intentFilter);
    }

    public void removeListener() {
        this.listener = BecomingNoisyListener.NO_OP;
        try {
            context.unregisterReceiver(this);
        } catch (Exception ignore) {
            // ignore if already unregistered
        }
    }
}

File: android/src/main/java/com/brentvatne/receiver/AudioBecomingNoisyReceiver.kt
package com.brentvatne.receiver

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.media.AudioManager
import androidx.core.content.ContextCompat

class AudioBecomingNoisyReceiver(private val context: Context) : BroadcastReceiver() {
    private var listener: BecomingNoisyListener = BecomingNoisyListener.NO_OP
    private val appContext: Context = context.applicationContext

    override fun onReceive(context: Context?, intent: Intent?) {
        if (intent != null && AudioManager.ACTION_AUDIO_BECOMING_NOISY == intent.action) {
            listener.onAudioBecomingNoisy()
        }
    }

    fun setListener(listener: BecomingNoisyListener) {
        this.listener = listener
        val intentFilter = IntentFilter(AudioManager.ACTION_AUDIO_BECOMING_NOISY)
        try {
            ContextCompat.registerReceiver(appContext, this, intentFilter, ContextCompat.RECEIVER_NOT_EXPORTED)
        } catch (e: Exception) {
            // Handle the exception and log it, but do not crash the app
            e.printStackTrace()
        }
    }

    fun removeListener() {
        this.listener = BecomingNoisyListener.NO_OP
        try {
            appContext.unregisterReceiver(this)
        } catch (ignore: Exception) {
            // ignore if already unregistered
        }
    }
}

interface BecomingNoisyListener {
    fun onAudioBecomingNoisy()

    companion object {
        val NO_OP: BecomingNoisyListener = object : BecomingNoisyListener {
            override fun onAudioBecomingNoisy() {
                // No operation
            }
        }
    }
}
